inference-engine/src/transformations/src/transformations/op_conversions/einsum_decomposition.cpp

```diff
@@ -0,0 +1,683 @@
+    const std::string& output_subscript, size_t input_ind, ngraph::NodeVector& subgraph_nodes) {
+    // perform sanity check for arguments
+    auto num_inputs = input_nodes.size();
+    NGRAPH_CHECK(num_inputs == input_subscripts.size(), "Each input must have own subscript.");
+    NGRAPH_CHECK(input_ind < num_inputs, "Input index is out of range.");
+
+    std::vector<int64_t> reduced_axes;
+    auto labels = ngraph::opset7::Einsum::extract_labels(input_subscripts[input_ind]);
+    std::string new_input_subscript = "";
+    for (size_t dim_ind = 0; dim_ind < labels.size(); ++dim_ind) {
+        const auto& label = labels[dim_ind];
+
+        // check if the current label is met in the other input subscripts
+        // or the output subscript
+        bool is_dim_reduced = is_dimension_reduced(input_subscripts, output_subscript, label, {input_ind});
+
+        // if label is not met, dimension corresponding to the label is to reduce
+        if (is_dim_reduced) {
+            reduced_axes.push_back(dim_ind);
+        } else {
+            new_input_subscript += label;
+        }
+    }
+
+    if (reduced_axes.size() == 0) {
+        // there is no axis to reduce
+        return;
+    }
+
+    // reduce by summed up elements along dimension for which label is met just once
+    const auto& input_node = input_nodes[input_ind];
+    auto axes_const = ngraph::opset7::Constant::create(ngraph::element::Type_t::i64, ngraph::Shape {reduced_axes.size()}, reduced_axes);
+    auto reduce_sum = einsum_decompose_ptr->register_new_node<ngraph::opset7::ReduceSum>(input_node, axes_const, false);
+
+    // update a vector of inputs and input subscripts
+    input_nodes[input_ind] = reduce_sum->output(0);
+    input_subscripts[input_ind] = new_input_subscript;
+
+    // update a vector of nodes for copy_runtime_info
+    subgraph_nodes.insert(subgraph_nodes.end(), {axes_const, reduce_sum});
+}
+
+/// \brief      Contract two inputs of Einsum operation according to equation.
+/// The result of the contraction is appended into input_nodes along with its subscript.
+/// The input nodes for these two operands are removed from input_nodes along with their input
+/// subscripts
+///
+/// \param      einsum_decompose_ptr    A pointer to Einsum decomposing pass
+/// \param      input_nodes             A vector of input nodes to Einsum operation
+/// \param      input_subscripts        A vector of corresponding subscripts for the input nodes
+/// \param      output_subscript        The output subscript
+/// \param      input_ind1              An index of the first operand
+/// \param      input_ind2              An index of the second operand
+/// \param      subgraph_nodes          A vector of operation nodes that is included into a
+/// sub-graph decomposing Einsum that is needed for copy_runtime_info
+///
+void contract_two_inputs(ngraph::pass::EinsumDecomposition* einsum_decompose_ptr,
```

Comment:
If we need to have a pointer to pass, will be better to have this function as a private member of transformation pass.

Reply:
We agreed to fix it in the scope of another PR for reference Einsum Implementation